aea/aea.py

```diff
@@ -0,0 +1,290 @@
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This module contains the implementation of an Autonomous Economic Agent."""
+import base64
+import importlib.util
+import inspect
+import logging
+import os
+import re
+from abc import abstractmethod, ABC
+from typing import Optional, Dict
+
+from aea.protocols.base.abstract_handler import AbstractHandler
+
+from aea.agent import Agent
+from aea.mail.base import Envelope, ProtocolId
+from aea.protocols.base.protocol import Protocol
+from aea.protocols.default.message import DefaultMessage
+from aea.protocols.default.serialization import DefaultSerializer
+
+logger = logging.getLogger(__name__)
+
+SkillId = str
+
+
+class Registry(ABC):
+    """This class implements an abstract registry."""
+
+    @abstractmethod
+    def populate(self, directory: str) -> None:
+        """
+        Load into the registry as specified in the config and apply consistency checks.
+
+        :param directory: the filepath to the agent's resource directory.
+        :return: None
+        """
+
+    @abstractmethod
+    def teardown(self) -> None:
+        """
+        Teardown the registry.
+
+        :return: None
+        """
+
+
+class ProtocolRegistry(Registry):
+    """This class implements the handlers registry."""
+
+    def __init__(self) -> None:
+        """
+        Instantiate the registry.
+
+        :return: None
+        """
+        self._protocols = {}  # type: Dict[ProtocolId, Protocol]
+        self._handlers = {}  # type: Dict[ProtocolId, AbstractHandler]
+
+    def populate(self, directory: str) -> None:
+        """
+        Load the handlers as specified in the config and apply consistency checks.
+
+        :param directory: the filepath to the agent's resource directory.
+        :return: None
+        """
+
+        protocols_spec = importlib.util.find_spec(".".join([directory, "protocols"]))
+        protocols_packages = list(filter(lambda x: not x.startswith("__"), protocols_spec.loader.contents()))
+        logger.debug("Processing the following protocol packages: {}".format(protocols_packages))
+        for protocol_name in protocols_packages:
+            try:
+                self._add_protocol(directory, protocol_name)
+            except Exception:
+                logger.exception("Not able to add protocol {}.".format(protocol_name))
```

Comment:
After adding we should probably quickly test whether each protocol implements all required methods?